plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,971 @@
+	guint chunks_sz = 0x0;
+
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x00, DDC_DATA_PAGE_SIZE);
+	chunks_sz = fu_chunk_array_length(chunks);
+	for (gint retry = 1; retry <= DDC_RW_MAX_RETRY_CNT; retry++) {
+		g_clear_error(&error_local);
+		for (guint i = 0; i < chunks_sz; i++) {
+			g_autoptr(FuChunk) chk = fu_chunk_array_index(chunks, i);
+
+			/* fast forward if chunk is empty, on the contrary set data per fragment
+			 * size */
+			if (fu_mediatek_scaler_device_chunk_data_is_blank(chk)) {
+				if (!fu_mediatek_scaler_device_set_data_fast_forward(self,
+										     chk,
+										     error))
+					return FALSE;
+			} else {
+				if (!fu_mediatek_scaler_device_set_data(self, chk, error))
+					return FALSE;
+				fu_device_sleep(FU_DEVICE(self), 1);
+			}
+
+			/* update progress */
+			fu_progress_set_percentage_full(fu_progress_get_child(progress),
+							(gsize)i + 1,
+							(gsize)chunks_sz);
+		}
+
+		/* exit the try loop when successes */
+		fu_device_sleep(FU_DEVICE(self), FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS);
+		if (fu_mediatek_scaler_device_check_sent_info(self, fw, &error_local))
+			return TRUE;
+
+		if (!g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_WRITE)) {
+			g_propagate_error(error, g_steal_pointer(&error_local));
+			return FALSE;
+		}
+
+		g_debug("retry write_firmware: step: %d, max: %d", retry, DDC_RW_MAX_RETRY_CNT);
+	}
+	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "maximum tries exceeded");
+	return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	guint max_tries = 30;
+
+	/* open device */
+	locker = fu_device_locker_new(self->i2c_dev, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* set the target address */
+	if (!fu_mediatek_scaler_ensure_device_address(self,
+						      FU_DDC_I2C_ADDR_DISPLAY_DEVICE >> 1,
+						      error))
+		return FALSE;
+
+	/* reboot the device */
+	if (!(fu_mediatek_scaler_device_set_isp_reboot(self, error)))
+		return FALSE;
+
+	/* wait for the device back */
+	if (!fu_device_retry_full(device,
+				  fu_mediatek_scaler_display_is_connected_cb,
+				  max_tries,
+				  FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL,
+				  NULL,
+				  error)) {
+		g_prefix_error(error,
+			       "display controller did not reconnect after %u retries: ",
+			       max_tries);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_write_firmware(FuDevice *device,
+					 FuFirmware *firmware,
+					 FuProgress *progress,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	gsize fw_size = 0;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```

Comment:
it would be good to the the actual numbers from the daemon and to remove GUESSED -- they're output in red text so they're kinda hard to miss :) Just do a new fixup commit (don't rebase) and we can squash all together when ready to merge.

Reply:
done

Reply:
Ohh, so the values were set correctly before?

Reply:
If so, please rebase against master and squash all the commits down into one commit and then LGTM. Thanks.

Reply:
> Ohh, so the values were set correctly before?

yep, I've modified the value to respect the calculation result.

Reply:
rebased on main and squashed the commits, thanks @hughsie .